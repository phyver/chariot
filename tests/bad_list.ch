-- simple terminating infinite list using lazy product type

codata prod_2('a,'b) where
    Fst : prod_2('a,'b) -> 'a
  | Snd : prod_2('a,'b) -> 'b

data list('a) where
    Nil : list('a)
  | Cons : 'a * list('a) -> list('a)

data nat where Zero : nat | Succ : nat -> nat

val bad : nat -> nat * list(nat)
  | (bad n).Fst = n
  | (bad n).Snd = Cons (bad (Succ n))

val inf : list(nat)
  | inf = Cons (bad 0)

:show bad
