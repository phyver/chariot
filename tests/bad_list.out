!! warning: the definition bad is NOT provably correct (weight_bound: 2, depth_bound: 2)
val
   bad : nat → nat * list(nat)
    | (bad n).Fst² = n
    | (bad n).Snd² = Cons³ (bad (n+1))

